akka-stream/src/main/scala/akka/stream/javadsl/Keep.scala

```diff
@@ -3,18 +3,51 @@
 
+import java.util.concurrent.{ CompletableFuture, CompletionStage }
+
 import akka.NotUsed
-import akka.japi.function
+import akka.japi.function.Function2
 import akka.japi.Pair
 
 object Keep {
-  private val _left = new function.Function2[Any, Any, Any] with ((Any, Any) ⇒ Any) { def apply(l: Any, r: Any) = l }
-  private val _right = new function.Function2[Any, Any, Any] with ((Any, Any) ⇒ Any) { def apply(l: Any, r: Any) = r }
-  private val _both = new function.Function2[Any, Any, Any] with ((Any, Any) ⇒ Any) { def apply(l: Any, r: Any) = new akka.japi.Pair(l, r) }
-  private val _none = new function.Function2[Any, Any, NotUsed] with ((Any, Any) ⇒ NotUsed) { def apply(l: Any, r: Any) = NotUsed }
-
-  def left[L, R]: function.Function2[L, R, L] = _left.asInstanceOf[function.Function2[L, R, L]]
-  def right[L, R]: function.Function2[L, R, R] = _right.asInstanceOf[function.Function2[L, R, R]]
-  def both[L, R]: function.Function2[L, R, L Pair R] = _both.asInstanceOf[function.Function2[L, R, L Pair R]]
-  def none[L, R]: function.Function2[L, R, NotUsed] = _none.asInstanceOf[function.Function2[L, R, NotUsed]]
+  def completionStage[T](value: T): CompletionStage[T] = {
+    val future = new CompletableFuture[T]()
+    future.complete(value)
+    future
+  }
+
+  private val _left = Function2[Any, Any, Any] { (l: Any, _: Any) ⇒ l }
+  private val _right = Function2[Any, Any, Any] { (_: Any, r: Any) ⇒ r }
+  private val _both = Function2[Any, Any, Any] { (l: Any, r: Any) ⇒ Pair(l, r) }
+  private val _none = Function2[Any, Any, NotUsed] { (_: Any, _: Any) ⇒ NotUsed }
+
+  private val _leftAsync = Function2[Any, Any, CompletionStage[Any]] {
+    (l: Any, _: Any) ⇒ completionStage(l)
+  }
+  private val _rightAsync = Function2[Any, Any, CompletionStage[Any]] {
+    (_: Any, r: Any) ⇒ completionStage(r)
+  }
+  private val _bothAsync = Function2[Any, Any, CompletionStage[Any]] {
+    (l: Any, r: Any) ⇒ completionStage(Pair(l, r))
+  }
+  private val _noneAsync = Function2[Any, Any, CompletionStage[NotUsed]] {
+    (_: Any, _: Any) ⇒ completionStage(NotUsed)
+  }
+
+  def left[L, R]: Function2[L, R, L] = _left.asInstanceOf[Function2[L, R, L]]
+  def right[L, R]: Function2[L, R, R] = _right.asInstanceOf[Function2[L, R, R]]
+  def both[L, R]: Function2[L, R, L Pair R] = _both.asInstanceOf[Function2[L, R, L Pair R]]
+  def none[L, R]: Function2[L, R, NotUsed] = _none.asInstanceOf[Function2[L, R, NotUsed]]
+
+  def leftAsync[L, R]: Function2[L, R, CompletionStage[L]] =
+    _leftAsync.asInstanceOf[Function2[L, R, CompletionStage[L]]]
+
+  def rightAsync[L, R]: Function2[L, R, CompletionStage[R]] =
+    _rightAsync.asInstanceOf[Function2[L, R, CompletionStage[R]]]
+
+  def bothAsync[L, R]: Function2[L, R, CompletionStage[L Pair R]] =
+    _bothAsync.asInstanceOf[Function2[L, R, CompletionStage[L Pair R]]]
+
+  def noneAsync[L, R]: Function2[L, R, CompletionStage[NotUsed]] =
+    _noneAsync.asInstanceOf[Function2[L, R, CompletionStage[NotUsed]]]
```

Comment:
I don't think that this belongs in this PR at all, please provide a ticket describing why this is needed and then after some discussion a separate PR containing this and the stuff in `japi.function.Function`